{% extends "home/base.html" %}
{% load static %}
{% block addon_css %}
<link rel="stylesheet" href="{% static 'home/css/jquery.dataTables.min.css' %}" type="text/css" />
<link rel="stylesheet" href="{% static 'home/css/structure_browser.css' %}" type="text/css" />
<link href="{% static 'home/css/residue_table.css' %}" rel="stylesheet">
<link href="{% static 'home/css/alignment.css' %}" rel="stylesheet">
<style>
.tooltip-inner {
    white-space: pre-wrap;
}
</style>
{% endblock %}

{% block content %}
<h3>Estimated economic burden on the NHS of GPCR targeting drugs</h3><br>
<div id="burden" class="chart_container" style="display: none;">
       <div id="burden"><svg id="burden" style="width:1050;height:500px;"></svg></div>

<div class="row">
<div class="col-sm-3"></div>
  <div class="col-md-6">
     <button id="update">add scaling factor for drugs targeting polymorphic putative functional sites</button>
     <input id="updater" type="text" maxlength="5" size="6" value="1.0" name="cutoff"
  </div>
  <div class="col-sm-3"></div>
</div>
<br><br><br>
<h4>Calculations</h4>
<p>The economic burden was estimated by:</p>
<p style="text-align: center;"><em>estimated economic burden per drug (&pound;) <strong>=</strong> NHS cost per drug per year (&pound;) <strong>x</strong></em></p>
<p style="text-align: center;"><em>% individuals with a  MV in a functional site of the respective drug targets</em></p>
<!-- <p style="text-align: center;"><em>correction or scaling factor&nbsp;</em></p> -->
<p>where:</p>
<ul style="list-style-type: circle;">
<li>The average NHS cost is the average yearly cost over a 4-year period (2013-2016) per GPCR drug that is listed (2012 and 2017 have uncomplete sales data).</li>
<li>% Individuals is the percentage of affected individuals with a missense variant in a functional site of the respective drug target(s) (2,504 individuals from 1,000 Genomes Project Phase 3 genotype data)</li>
<li>The fraction of affected individuals was calculated using two different criteria (a) considering only individuals who have a variation in a functional site in both alleles (homozygous), which is more conservative and (b) individuals if they have a variation in a functional site in at least one allele (i.e. homozygous and heterozygous), which is less conservative.</li>
<li>Functional sites were distinguished into (a) known functional sites that included ligand binding, effector binding, PTM site, sodium binding site and microswitches (more conservative) and (b) putative functional site as those predicted to be deleterious based on SIFT or PolyPhen (less conservative).</li>
</ul>
<p>More specifically, for each drug we collected the respective targets and computed economic burden using the following four categories: i) individuals with homozygous alleles in known functional sites ii) individuals with at least one variant allele in a known functional site, iii) individuals with homozygous alleles in known or putative functional sites iv) individuals with at least one variant allele in a known or putative functional sites. For these estimates, we consider that each prescription (NHS data) is made for a unique individual, due to patient anonymity.
</p>
<br><br>
<h3>Sales and variation information per drug</h3><br>
    <!-- Download button -->
    <a class="btn btn-sm btn-primary" href="download" role="button">
    <span class="glyphicon glyphicon-download"></span> Download</a>
<div class="row">
    <div class="col-md-11 text-center">
        <div style="padding-top: 10px; font-size: 11px;white-space: nowrap;">
            <table width="100%" id="drugdata">
                <thead>
                    <tr>
                        <th style="text-align:center" class="citation-tooltip" target="_blank" data-toggle="tooltip"  data-container="body" data-html="true" data-original-title="Name of active ingredient" data-placement="top">Drug Name</th>
                        <th style="text-align:center" class="citation-tooltip" target="_blank" data-toggle="tooltip"  data-container="body" data-html="true" data-original-title="GPCR Protein Name" data-placement="top">GPCR Drug Targets</th>
                        <th style="text-align:center" class="citation-tooltip" target="_blank" data-toggle="tooltip"  data-container="body" data-html="true" data-original-title="average NHS sales from 2013-2016" data-placement="top">Sales Average in &pound;</th>
                        <th style="text-align:center" class="citation-tooltip" target="_blank" data-toggle="tooltip"  data-container="body" data-html="true" data-original-title="average drug prescriptions from 2013-2016" data-placement="top">Average number of prescriptions</th>
                        <th style="text-align:center" class="citation-tooltip" target="_blank" data-toggle="tooltip"  data-container="body" data-html="true" data-original-title="average sales devided by average prescription to estimate item/prescription costs" data-placement="top">Item Cost in &pound;</th>
                        <th style="text-align:center" class="citation-tooltip" target="_blank" data-toggle="tooltip"  data-container="body" data-html="true" data-original-title="no. of total variants in putative functional sites (predicted deleterious from SIFT or PolyPhen)" data-placement="top">No. of Putative Functional Site Variants</th>
                        <th style="text-align:center" class="citation-tooltip" target="_blank" data-toggle="tooltip"  data-container="body" data-html="true" data-original-title="no. of total variants in known functional site including ligand-binding sites, effector-binding sites, sodium ion pocket, microswitches and post-translational modification sites" data-placement="top">No. of Known Functional Site Variants</th>
                        <th style="text-align:center" class="citation-tooltip" target="_blank" data-toggle="tooltip"  data-container="body" data-html="true" data-original-title="Indication category after the British National Formulary (BNF)" data-placement="top">BNF section</th>
                      </tr>
                      <tr>
                          <th style="text-align:center"></th>
                          <th style="text-align:center"></th>
                          <th style="text-align:center"></th>
                          <th style="text-align:center"></th>
                          <th style="text-align:center"></th>
                          <th style="text-align:center"></th>
                          <th style="text-align:center"></th>
                          <th style="text-align:center"></th>
                      </tr>
                      <tr>
                        <th style="text-align:center"></th>
                        <th style="text-align:center"></th>
                        <th style="text-align:center"></th>
                        <th style="text-align:center"></th>
                        <th style="text-align:center"></th>
                        <th style="text-align:center"></th>
                        <th style="text-align:center"></th>
                        <th style="text-align:center"></th>
                      </tr>
                </thead>
                <tfoot>
                    <tr>
                      <th style="text-align:center">Drug Name</th>
                      <th style="text-align:center">GPCR Drug Targets</th>
                      <th style="text-align:center">Sales Average in &pound;</th>
                      <th style="text-align:center">Average number of prescriptions</th>
                      <th style="text-align:center">Item Cost in &pound;</th>
                      <th style="text-align:center">Known Functional Site Variants</th>
                      <th style="text-align:center">Putative Functional Site Variants</th>
                      <th style="text-align:center">BNF section</th>
                    </tr>
                </tfoot>
                {% for drug in drug_data %}
                <!-- add: .select_related.all -->
                <tr>
                  <td style="text-align:center"><a href='/drugs/nhs/{{drug.drugname|safe}} '> {{drug.drugname|safe}}</a></td>
                  <td>{% for target in drug.targets %}
                  <a href='/mutational_landscape/protein/{{ target|add:"_human" }} '>{{ target }}</a>
                  {% endfor %}</td>
                  <td>{{drug.average_cost}}</td>
                  <td>{{drug.average_items}}</td>
                  <td>{{drug.item_cost}}</td>
                  <td>{{drug.putative_func}}</td>
                  <td>{{drug.known_func}}</td>
                  <td>{% for section in drug.section %}
                  <a href='/drugs/nhs/section/{{ section }} '>{{ section }}</a>
                  {% endfor %}</td>
                </tr>
                {% endfor %}
            </table>
        </div>
      </div>
    </div>
<br><br>
{% endblock %}
{% block addon_js %}
<script src="{% static 'home/js/saveSvgAsPng.js' %}"></script>
<script src="{% static 'home/js/d3.min.js' %}"></script>
<script src="{% static 'home/js/d3.tip.js' %}"></script>
<script src="{% static 'home/js/nv.d3.min.js' %}"></script>
<script src="{% static 'home/js/jquery.dataTables.min.js' %}"> </script>
<script src="{% static 'home/js/dataTables.tableTools.min.js' %}"> </script>
<script src="{% static 'home/js/jquery.dataTables.columnFilter.js' %}"> </script>
<script src="{% static 'home/js/alignment.js' %}"></script>

<script type="text/javascript" charset="utf-8">
        $(document).ready(function() {
            var table = $('#drugdata').DataTable({
                // 'scrollX': true,
                /*// 'scrollY': $(window).height()-300,*/
                // 'scrollCollapse': true,
                'paging': true,
                'lengthMenu': [[20, 50, 100, -1], [20, 50, 100, "All"]],
                'orderCellsTop': true,
                "bInfo" : false,
                'autoWidth': false,
                // 'sScrollX': "100%",
                // 'sScrollXInner': "100%",
                'order': [[3, "desc"]],
                'dom': 'T<"clear">lfrtip',
                'columnDefs': [
                  {"width": "10%", targets: 0},
                  {"width": "20%", targets: 1},
                  {"width": "10%", targets: 2},
                  {"width": "10%", targets: 3},
                  {"width": "30%", targets: 4},
                  {"width": "10%", targets: 5},
                  {"width": "10%", targets: 6},
                  {"width": "5%", targets: 7}],
                'tableTools': {
                    "sRowSelect": "multi",
                    "aButtons": []
                },
                initComplete: function () {
                    $('#drugdata').dataTable().columnFilter({
                        'sPlaceHolder': "head:after",
                        'aoColumns': [
                            { type: "text" },
                            { type: "text" },
                            { type: "text" },
                            { type: "text" },
                            { type: "text" },
                            { type: "text" },
                            { type: "text" },
                            { type: "text" },
                        ]
                    });
                }

            });
        });
    </script>
<script>
// https://github.com/novus/nvd3/tree/master/examples
var chart ;
var chartData;

window.onload = function () {

  nv.addGraph(function() {
      var datum = {{data|safe}}

      chart = nv.models.multiBarChart()
        .reduceXTicks(false)   //If 'false', every single x-axis tick label will be rendered.
        // .rotateLabels(0)      //Angle to rotate x-axis labels.
        .showControls(true)   //Allow user to switch between 'Grouped' and 'Stacked' mode.
        // .groupSpacing(0.1)    //Distance between each group of bars.
        .stacked(true)
        .margin({"left": 120, "right": 50, "top": 120, "bottom": 70,})
      ;

      chart.xAxis
          // .axisLabel("correction or scaling factor")
          // .showMaxMin(true)
          // .tickFormat(d3.format(',.1f'))

      chart.yAxis
          .axisLabel("estimated economic burden per year (GBP, £)")
          .axisLabelDistance(35)
          .tickFormat(d3.format(','));

      chart.tooltip.contentGenerator(function (d) {
          var html = "<h3>"+d.value+"</h3> <ul>";

          d.series.forEach(function(elem){
            html += "<h4 style='color:"+elem.color+"'>"
                    +elem.key+"</h4> estimated economic burden: <b>"+elem.value.toFixed(2).replace(/(\d)(?=(\d{3})+\.)/g, '$1,')+" &pound  "+"</b>";
          })
          html += "</ul>"
          return html;
        })

      chart.multibar.dispatch.on("elementClick", function(e) {
        link = e['data']['key']
          window.open ('/drugs/nhs/section/'+link,'_self',false)
      })

      chartData = d3.select('#burden svg').datum(datum)
      chartData.transition().duration(500).call(chart);

      nv.utils.windowResize(chart.update);

      return chart;
  });

  var raw = {{data|safe}};
  function update(cf) {
    var cf = document.getElementById('updater').value

    var rawLength = raw.length;
    for (var i = 0; i < rawLength; i++) {

      var kk = raw[i]['values'].length;
      raw[i]['values'][kk] = {'y': raw[i]['values'][3]['y']*cf, 'x': cf}
    }

    // Update the SVG with the new data and call chart
    chartData.datum(raw).transition().duration(5).call(chart);
    nv.utils.windowResize(chart.update);
};

d3.select("#update").on("click", update);

    /* Inspired by Lee Byron's test data generator. */
    function stream_layers(n, m, o) {
      if (arguments.length < 3) o = 0;
      function bump(a) {
        var x = 1 / (.1 + Math.random()),
            y = 2 * Math.random() - .5,
            z = 10 / (.1 + Math.random());
        for (var i = 0; i < m; i++) {
          var w = (i / m - y) * z;
          a[i] += x * Math.exp(-w * w);
        }
      }
      return d3.range(n).map(function() {
          var a = [], i;
          for (i = 0; i < m; i++) a[i] = o + o * Math.random();
          for (i = 0; i < 5; i++) bump(a);
          return a.map(stream_index);
        });
    }

    /* Another layer generator using gamma distributions. */
    function stream_waves(n, m) {
      return d3.range(n).map(function(i) {
        return d3.range(m).map(function(j) {
            var x = 20 * j / m - i / 3;
            return 2 * x * Math.exp(-.5 * x);
          }).map(stream_index);
        });
    }

    function stream_index(d, i) {
      return {x: i, y: Math.max(0, d)};
    }
  //Generate some nice data.
  function exampleData(sf) {
    return stream_layers(10,sf,0.1).map(function(data, i) {
      return {
        key: 'Stream #' + i,
        values: data
      };
    });
  }

  $(document).ready(function () {
      $('#burden').css("fill", '#000000');
      $('#burden').css("display", '');
      // draw('1');

  });
};
</script

{% endblock %}
